src/inmanta/deploy/tasks.py

```diff
@@ -87,81 +96,57 @@ class PoisonPill(Task):
     It functions mostly as a no-op
     """
 
-    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+    async def execute(self, task_manager: "scheduler.TaskManager", agent: str) -> None:
         pass
 
 
-class OnLatestState(Task):
-
-    async def execute(self, scheduler: "scheduler.ResourceScheduler", agent: str) -> None:
+class Deploy(Task):
+    async def execute(self, task_manager: "scheduler.TaskManager", agent: str) -> None:
+        version: int
         resource_details: "state.ResourceDetails"
-        async with scheduler._scheduler_lock:
-            # fetch resource details atomically under lock
-            try:
-                resource_details = scheduler._state.resources[self.resource]
-            except KeyError:
-                # Stale resource, can simply be dropped.
-                # May occur in rare races between new_version and acquiring the lock we're under here. This race is safe
-                # because of this check, and an intrinsic part of the locking design because it's preferred over wider
-                # locking for performance reasons.
-                return
-        await self.execute_on_resource(scheduler, agent, resource_details)
-
-    @abc.abstractmethod
-    async def execute_on_resource(
-        self, scheduler: "scheduler.ResourceScheduler", agent: str, resource_details: "state.ResourceDetails"
-    ) -> None:
-        pass
+        intent = await task_manager.get_resource_intent(self.resource, for_deploy=True)
+        if intent is None:
+            # Stale resource, can simply be dropped.
+            return
```

Comment:
The common part here became so trivial that I felt it made no sense anymore to group them together under a common subclass, so I dropped `OnLatestState`.